list.lisp

```diff
@@ -0,0 +1,116 @@
+;;;; -*- Mode: Lisp; indent-tabs-mode: nil -*-
+
+(in-package :split-sequence)
+
+(declaim (inline
+          collect-until count-while
+          split-list split-list-if split-list-if-not
+          split-list-from-end split-list-from-start split-list-internal))
+
+(declaim (ftype (function (&rest t) (values list unsigned-byte))
+                split-list split-list-if split-list-if-not))
+
+(declaim (ftype (function (function list unsigned-byte (or null unsigned-byte) (or null unsigned-byte)
+                                    boolean)
+                          (values list unsigned-byte))
+                split-list-from-start split-list-from-end split-list-internal))
+
+(defun collect-until (predicate list end)
+  "Collect elements from LIST until one that satisfies PREDICATE is found.
+
+  At most END elements will be examined. If END is null, all elements will be examined.
+
+  Returns four values:
+
+  * The collected items.
+  * The remaining items.
+  * The number of elements examined.
+  * Whether the search ended by running off the end, instead of by finding a delimiter."
+  (let ((examined 0)
+        (found nil))
+    (flet ((examine (value)
+             (incf examined)
+             (setf found (funcall predicate value))))
+      (loop :for (value . remaining) :on list
+            :until (eql examined end)
+            :until (examine value)
+            :collect value :into result
+            :finally (return (values result
+                                     remaining
+                                     examined
+                                     (and (not found)
+                                          (or (null end)
+                                              (= end examined)))))))))
+
+(defun count-while (predicate list end)
+  "Count the number of elements satisfying PREDICATE at the beginning of LIST.
+
+  At most END elements will be counted. If END is null, all elements will be examined."
+  (if end
+      (loop :for value :in list
+            :for i :below end
+            :while (funcall predicate value)
+            :summing 1)
+      (loop :for value :in list
+            :while (funcall predicate value)
+            :summing 1)))
+
+(defun split-list-internal (predicate list start end count remove-empty-subseqs)
+  (let ((count count)
```

Comment:
why (count count) ?

Reply:
I mutate `COUNT` in line 66 and I do not want to modify the original argument that the function was invoked with - this is for debugging purposes.

Reply:
Right, makes sense. I usually would prefer to wrap all execution variables in a struct and define separate top-level functions. It is somewhat difficult to mentally keep track of all those pesky variables, especially when local functions do multiple things at once. Which is the case here, in both functions actually. Gather-chunk does not only gather-chunk but also mutates list binding. Should-collect-p also mutates binding of count. This is not ideal IMHO.

Perhaps restructuring this into more functional style would be better. You can write higher order function called reduce-chunks that would handle constructing chunks and driving iteration forward while keeping it separate from actually gathering chunks. In fact such function would be useful on it's own merrits.

Reply:
I am unsure if I am able to do that myself - the list implementation I originally wrote ended up getting rewritten by @sjl anyway due to its lack of readability, and his current code is already one league above my original.

I wonder how a more functional approach would do in case of performance though, since we want to keep this code as performant as possible but still legible.

Reply:
First of, the current code is not far from becoming reduce-chunks. Simply separate all crap that actually drives iteration forward (that would be gather-chunk, mostly) into a separate function. This function would accept callback that would accept two arguments (previous state and next chunk) and would produce next state. Calling reduce-chunks with something that would sometimes cons chunk into list is pretty much close to what you are trying to get. Secondly, performance should be just fine and perhaps even identical to your current version if reduce-chunks would be declaimed as inline.

Reply:
Keep in mind that making this code functionally equivalent to the old `SPLIT-SEQUENCE` with regards to all outputs severely limits the amount of freedom that refactoring can do; me and @sjl have fell into many such traps while implementing this list code, and we've both started with very simple and neat codebases that then were hammered into compliance with the test suite.